### ISSUE TITLE
Interface ThingSpeak with BB-WI

### ISSUE THREAD

**Issue by zrajani** [zrajani/iLETTERBOX_CAPSTONE_PROJECT_EMBT_F2020#26]
As for the last part of the software interface required for the project, we have to connect and send the data collected from the project to a cloud service and we choose this cloud service to be ThingSpeak and all the issues faced shall be illustrated in this issue thread.

**Comment by zrajani** [comment]
Connecting the Beagle bone wireless to the internet was a problem since it was unable to get connected to the cloud. The reason for which could be due to the setup was far away from the modem that is the setup was made in the basement of the apartment and a similar scenario was observed during issue #6 thus we had to figure a way connect the board to the wireless network

**Comment by zrajani** [comment]
> Connecting the Beagle bone wireless to the internet was a problem since it was unable to get connected to the cloud. The reason for which could be due to the setup was far away from the modem that is the setup was made in the basement of the apartment and a similar scenario was observed during issue #6 thus we had to figure a way connect the board to the wireless network

Inorder to resolve this issue we had two methods in mind which were as follows:

- [ ] 1. To move the setup closer to the modem or wireless network. 

- [ ] 2.  Find another suitable network to connect in the current location 

- [x] 3. Other (which was strange)

Thus for us we just had to clear some obstacles from the way like the basement door of the apartment was closed but when the basement door was kept open and the setup was tried to connect to the internet it started to work with an acceptable internet speed. As connecting to the mobile hotspot somewhere is not possible as that network is not been shown on the BB-WI and the first option since we have no permission to get closer to the modem (personal terms with the place I stay) thus the option was not feasible. 
But somehow strangely the issue was resolved by just opening the door and may be having an obstacle-free for the radio waves which helps us connect to the wireless network.

**Comment by zrajani** [comment]
The next issue faced was reading the `using namespace exploringBB` as the code was adopted from [Derrek Molley](https://github.com/derekmolloy/exploringBB/tree/version2/chp11/thingSpeak) thus during the compilation of the code it showed us the error that the namespace was not found thus, to resolve the issue we had again 2 options:

- [ ] 1. Write the code by ourselves which was based on the SocketClient.h in the Derrek Molley Repository (Refer to the link above)

- [x] 2. Clone his repository and then try to give the path of the SocketClient.cpp during the compilation 

Thus we used the second approach, in this case, to make things simpler and we specified the path of the SocketClient.cpp during the compilation command. Thus during compilation our command was `g++ <cpp file name> <path of the SocketClient.cpp> -o <output_file_name> -liobb` thus the program was compiling successfully now and giving us the desired output on the console of the Beaglebone Wireless.

**Comment by zrajani** [comment]
Now since the program was working smoothly on the console of the Beaglebone wireless we had to face the next issue regarding the data being upload to the cloud. The data on the console showed that it was able to connect to the Thingspeak API server and send the data but when viewed on the ThingSpeak service (channel) the data was received but not stored anywhere on the cloud. Thus we had to check the code that we got from [Derrek Molley](https://github.com/derekmolloy/exploringBB/tree/version2/chp11/thingSpeak) line by line and see what could have gone wrong thus, we saw that all things were looking good and seem to be working as per our desired logic. Thus we just had a small mistake in spelling the word "Field" which was causing the problems as in ThingSpeak we don't have anything like "Feild". Thus changing the spelling correctly we were able to see the data that was appearing in the cloud in real-time.

**Comment by zrajani** [comment]
The major issue we faced was with using a string stream header file which is used for output string stream and input string stream which is used in order to give both numerical or string data to the cloud (sstream is not available in c code which is the main string structure for sending data to the cloud), so we had to use C++ code as the header file is not available in C programming language. Thus deviation of the entire code now shifted from C language to C++ language but the code that we wrote in C was preserved as many functions of C are compatible in C++ whereas vice-versa is not possible.

Also, the delay by default is set to 15 seconds and can't be reduced unless we have a paid version but this 15 seconds gap allows unnecessary overlapping of data. Thus from our end, we incorporated the 15 seconds delay in the code itself using the sleep(15) command.
